### PR TITLE
AbstractKKTRHS: second attempt 

### PR DESCRIPTION
**Pull request by frapac**
This is a new attempt at implementing the `AbstractKKTRHS` abstraction. Compared to the previous one, this PR uses a cleaner interface for the KKT right-hand-side, and update the implementation of iterative refinement to be more modular. 

In details, this PR
- adds `ReducedKKTVector` and `UnreducedKKTVector` structures 
- removes module `MadNLPRichardson` and replaces it with a proper structure `RichardsonIterator` 
- rewrites the iterative refinement algorithm to be more modular 
- add documentation for the linear solver's interface

### CODE REVIEW

**Comment by frapac** [comment]
Thanks for the feedback @sshin23 ! What was the complaint about the variable name?

**Comment by sshin23** [comment]
@frapac oops, sorry. I think `values` in `KKTVector` is not intuitive. How about `full`?

**Comment by frapac** [comment]
That's a good suggestion! I will update the name for both the attribute and the getter

File: Project.toml [hunk 1]
```diff
@@ -19,7 +21,6 @@ SuiteSparse = "4607b0f0-06f3-5cda-b6b1-a6196a1729e9"
 [compat]
 MINLPTests = "~0.5"
 MadNLPTests = "~0.2"
-MathOptInterface = "1"
```
Comment (sshin23) [inline Project.toml:22]: Why is `MOI` compat removed?

**Review by sshin23 — APPROVED** [review]
🎉

File: Project.toml (hunk 1 above)
Comment (sshin23) [reply Project.toml:22]: And `MadNLPGPU` and `MadNLPTests` don't need to be here, no?

**Pull request merged**